src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataCollectionSerializer.cs

```diff
@@ -161,17 +174,26 @@ public ODataCollectionSerializer(ODataSerializerProvider serializerProvider)
                     IEdmTypeReference actualType = writeContext.GetEdmType(item, item.GetType());
                     Contract.Assert(actualType != null);
 
-                    itemSerializer = itemSerializer ?? SerializerProvider.GetEdmTypeSerializer(actualType);
-                    if (itemSerializer == null)
+                    //For instance annotations we need to use complex serializer for complex type and not ODataResourceSerializer
+                    if (isForAnnotations && actualType.IsComplex())
                     {
-                        throw new SerializationException(
-                            Error.Format(SRResources.TypeCannotBeSerialized, actualType.FullName()));
+                        valueCollection.Add(
+                        new ODataResourceValueSerializer(SerializerProvider).CreateODataValue(item, actualType, writeContext).GetInnerValue());
```

Comment:
Note that the previous code only set the itemSerializer once for the enumerable, where-as your code checks/sets it each time.

Assuming we continue to believe that the itemSerializer will be the same for each item in the enumerable (which seems reasonable), then you should still only set it once for the loop.

If null, you would check isForAnnotations && actualType.IsComplex(), then set itemSerializer to new ODataResourceValueSerializer(SerializerProvider), or to the GetEdmTypeSerializer.

In any case, you should move valueCollection.Add(itemSerializer.CreateODataValue(item, actualType,writeContext).GetInnerValue()) outside the if/else -- it's logic is the same against whichever itemSerializer you previously identified.

Something like:

                  if (itemSerializer == null)
                  { 
                      //For instance annotations we need to use complex serializer for complex type and not ODataResourceSerializer
                      if (isForAnnotations && actualType.IsComplex())
                      {
                          itemSerializer = new ODataResourceValueSerializer(SerializerProvider);
                      }
                      else
                      {
                          itemSerializer = SerializerProvider.GetEdmTypeSerializer(actualType);
                          if (itemSerializer == null)
                          {
                              throw new SerializationException(
                                  Error.Format(SRResources.TypeCannotBeSerialized, actualType.FullName()));
                          }
                    }

                    // ODataCollectionWriter expects the individual elements in the collection to be the underlying
                    // values and not ODataValues.
                    valueCollection.Add(
                         itemSerializer.CreateODataValue(item, actualType, writeContext).GetInnerValue());